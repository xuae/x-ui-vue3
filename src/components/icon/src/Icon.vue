<template>
  <i v-bind="$attrs" class="x-icon" :style="styles">{{ name }}</i>
</template>

<script lang="ts">
import { computed, defineComponent, toRefs } from 'vue';

export default defineComponent({
  name: 'XIcon',
  props: {
    // 图标名称
    name: {
      type: String,
      required: true,
    },

    // 尺寸
    size: {
      type: [String, Number],
    },

    // 颜色
    color: {
      type: String,
    },
  },
  setup(props, { attrs, slots, emit }) {
    const { size, color } = toRefs(props);
    const styles = computed(() => {
      const style: { [key: string]: string } = {};
      if (size) {
        style.fontSize = `${size.value}px`;
      }
      if (color && color.value) {
        style.color = color.value;
      }
      if (attrs.onClick) {
        style.cursor = 'pointer';
      }
      return style;
    });

    return { styles };
  },
});
</script>
