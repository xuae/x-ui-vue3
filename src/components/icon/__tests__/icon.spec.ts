import { mount } from '@vue/test-utils';
import Icon from '@/components/icon/src/Icon.vue';

const name = 'name';

describe('Icon.vue', () => {
  it('create', async () => {
    const wrapper = mount(Icon, {
      props: {
        name,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.classes('x-icon')).toBe(true);
    expect(wrapper.text()).toBe(name);
  });

  test('prop color', async () => {
    const wrapper = mount(Icon, {
      props: {
        name,
        color: 'red',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.props('color')).toBe('red');
    expect(wrapper.attributes('style')).toContain('color: red');
    await wrapper.setProps({
      color: 'green',
    });
    expect(wrapper.attributes('style')).toContain('color: green');
  });

  it('prop size', async () => {
    const wrapper = mount(Icon, {
      props: {
        name,
        size: '36',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.props('size')).toBe('36');
    expect(wrapper.attributes('style')).toContain('font-size: 36px');
    await wrapper.setProps({
      size: 48,
    });
    expect(wrapper.attributes('style')).toContain('font-size: 48px');
  });

  it('prop type', async () => {
    const data = ['outlined', 'round', 'sharp', 'twoTone'];
    for (let i = 0; i < data.length; i++) {
      const wrapper = mount(Icon as any, {
        props: {
          name,
          type: data[i],
        },
      });
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.classes(`x-icon__${data[i]}`)).toBe(true);
    }
  });

  it('bind click', () => {
    const onClick = jest.fn();
    const wrapper = mount(Icon, {
      props: {
        name,
      },
      attrs: {
        onClick: onClick,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.trigger('click');
    expect(onClick).toHaveBeenCalled();
    expect(wrapper.attributes('style')).toContain('cursor: pointer');
  });
});
