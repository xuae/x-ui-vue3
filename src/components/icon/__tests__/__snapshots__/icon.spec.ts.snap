// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon.vue bind click 1`] = `<i class="x-icon" style="cursor: pointer;">name</i>`;

exports[`Icon.vue create 1`] = `<i class="x-icon">name</i>`;

exports[`Icon.vue prop color 1`] = `<i class="x-icon" style="color: red;">name</i>`;

exports[`Icon.vue prop size 1`] = `<i class="x-icon" style="font-size: 36px;">name</i>`;
